Upstream master bump 0513

jjsjann123 opened the pull request:
Updating nvfuser code base.

This should fix the indexing issue observed in https://github.com/pytorch/vision/issues/6015.

Running tests locally as well. Will update the description here at a later point

@bypass-github-export-checks

csarofeen committed:
Make dispatch of KIR and FusionIR more similar. (#1314)

csarofeen committed:
Fix test names. (#1329)

crcrpar committed:
add missing terminating " character (#1330)

rdspring1 committed:
Type Promotion Fixes (#1322)

* Allow cast from Int to Int32 type
* Update test_binary_ops with scalar tests
* Add integral scalars to optional cast exception list

csarofeen committed:
Implement a visitor like class for KIR, move passes to it. (#1332)

jjsjann123 committed:
fixing conv2d decomposition and tests (#1333)

csarofeen committed:
Create mutator class for kir and refactor passes (#1336)

csarofeen committed:
Refactor get allocation information in lower_utils (#1337)

... and reuse in lower_allocation pass.

jjsjann123 committed:
Alias copy patch (#1338)

fixes the assertion from #1325 on our devel branch.

1. update alias information after graph mutation
2. patch unsqueeze: i. support negative dimension; ii. fixing range check

kevinstephano committed:
Add rsub for functorch support. (#1342)

jjsjann123 committed:
Fixes patches from ltc aot autograd etc (#1340)

Fixing a few smaller issues here and there:

Exposing python API to switch single node fusion;
Exposing python API to switch horizontal fusion (Needed to avoid PW scheduler failure on fusion with outputs of different shapes/ranks);
Adding shape expression short-cut support for native_dropout (Bug reported by AOTAutograd);
Fixing device check to avoid fusion of node with inputs on different device. Long term we should have supported this, but disabling it for now to avoid assert. (e.g. scalar cpu tensor can be operated on cuda tensors, feature from TensorIterator).

jjsjann123 committed:
Nvfuser code bump 12 5 (#69964) (#1345)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/69964

Things added in this PR that requires review:
1. cuLaunchCooperativeKernel driver API added
aten/src/ATen/cuda/detail/LazyNVRTC.cpp
aten/src/ATen/cuda/nvrtc_stub/ATenNVRTC.h

nvfuser code update:
1. perf turning on codegen scheduler that improves performance.
2. permutation support has been extended beyond contiguous/channels-last. (The improvements could be observed on PW benchmark)

Things reverted from local changes:
1. aten::gelu with approximation
2. local changes that is upstreamed in PR https://github.com/pytorch/pytorch/issues/68804

Pull Request resolved: https://github.com/pytorch/pytorch/pull/69428

Reviewed By: ngimel

Differential Revision: D33073817

Pulled By: wconstab

fbshipit-source-id: e77d32e81d037d7370822b040456fd4c3bd68edb

csarofeen committed:
Fix segfault. (#1357)

csarofeen committed:
Collection of refactoring in nvFuser lowering (#1339)

* Refactor War Sync Insertion Pass (#1339)
* Remove kir::Expr::scope_ (#1341)
* Fusion IR Refactor (#1343)
* Refactor KIR Step 1 - Remove kir::Node (#1347)
* Refactor KIR Step 2 - TMP IrUtils change (#1348)
* Refactor KIR Step 3 - Remove kir::Expr and kir::Val. (#1349)
* Refactor KIR Step 4 - Remove kir::Bool,Double,Int,NamedScalar. (#1350)
* Refactor KIR Step 5 - Remove kir::IterDomain/TensorDomain/TensorView (#1351)
* Refactor KIR Step 6 - Remove 
 kir::UnaryOp/BinaryOp/TernaryOp/ReductionOp/WelfordOp/BroadcastOp. (#1352)
* Refactor KIR Step 7 - Remove kir dispatch (#1353)
* Refactor KIR Step 8 - Clean up lower_utils (#1355)
* Refactor KIR Step 9 - lower_utils ir_utils::applyReplacements. (#1354)
* Refactor KIR Step 10 - Remove kir_printer in favor of io_stream (#1356)

naoyam committed:
Clang format (#1360)

naoyam committed:
Support more flexible padding sizes in shift and gather  (#1334)

 This PR relaxes the constraint so that arbitrary padding sizes can be used as long as output domains don't get larger than input domains.

naoyam committed:
clang-tidy (#1363)

naoyam committed:
Print CA info only when FIR (#1364)

jjsjann123 committed:
Transposing scalar tensor patch (#1361)

naoyam committed:
Build error fix (and clang-format) (#1368)

rdspring1 committed:
Fixes #1310 - alias_copy assertion in fallback path (#1335)

* Implement alias_copy operations only for CudaFusionGroup to support fallback path
* Remove alias (a) annotation from alias_copy schema

shmsong committed:
Segment independent component on fusion graph (#1370)

* force segment un-connected graphs

* derive heuristic on empty groups

* add test

* lint

* handled aliased output in batchnorm

* empty tensor

* lint and comment

* clang format

* check reference tv available in pointwise scheduler

* comment

* cleanup test and check utils

csarofeen committed:
Avoid constructing a new TV in parsing. (#1374)

csarofeen committed:
Remove Kernel IR join infrastructure with Fusion (#1373)

* Have Kernel Inherit IrContainer (#1375)
* Kernel<-Fusion Step 1 - Convert ExprSort to StmtSort (#1376)
* Kernel<-Fusion Step 2 - Mutator refactor (#1377)
* Kernel<-Fusion Step 3 - Debug print for expr_eval and type promotion fix (#1379)
* Kernel<-Fusion Step 4 - Have kernel inherit Fusion (#1380)
* Kernel<-Fusion Step 5 - Move lowering passes into their own files (#1382)
* Kernel<-Fusion Step 6 - Remove kir::IrBuilder (#1383)
* Kernel<-Fusion Step 7 - Remove kir functions from ComputeAtMap (#1384)
* Kernel<-Fusion Step 8 - Clean up [lower/executor] utils (#1387)
* Kernel<-Fusion Step 9 - Remove TensorView::fuserTv (#1388)
* Kernel<-Fusion Step 10 - Remove lowerVal/lowerExpr (#1389)
* Kernel<-Fusion Step 11 - Finish cleaning up kir (#1390)

naoyam committed:
clang-format (#1394)

naoyam committed:
Pass inputs to compileFusion to avoid redundant compilation (#1395)

naoyam committed:
Double buffering support (#1381)

Adds TensorView::doubleBuffer(). See the new tests how it is used.

For an overview of the lowering algorithm, please see lower_double_buffer.h.

csarofeen committed:
Some minor fixes. (#1401)

jjsjann123 committed:
Graph lint error patch (#1378)

1. extend buildShapeExpression for squeeze_copy/unsqueeze_copy ops.
2. patching broadcastSizes insertion point for buildShapeExpression to avoid graph::copy() linter assert.
3. adding tests
4. supports no-op squeeze (squeezing on dimension that's not size-1)

TODO (in follow up PRs):
1. extend buildShapeExpression to view_copy and reshape_copy as well
2. refactor broadcastSizesExpression to allow graceful failure instead of hard assert

shmsong committed:
new clang-format binary hash (#1398)

naoyam committed:
Verify vectorization eligibility for intermediate tensors (#1402)

* Verify vectorization eligibility for intermediate tensors

csarofeen committed:
Add nullptr protection. (#1407)

jjsjann123 committed:
Merge commit '17540c5c80f5c6cd4e0fee42ec47d881e46f47f9' into HEAD

csarofeen committed:
Fix non-unrolled reduction scheduling. (#1409)

naoyam committed:
Detect non-concretized broadcast domains (#1412)

Used to avoid unnecessary parallel broadcast operations.

Needed to clean up the thread predicate info passed to Kernel. The only
necessary information is which parallel types are in fact parallel in
broadcast ops, so a map holding that information is added to
KernelSummary. ThreadPredicateMap is dropped from Kernel.


Co-authored-by: Christian Sarofeen <csarofeen@nvidia.com>

jjsjann123 committed:
support device promotion on scalar tensor in nvfuser (#1400)

Fixes #1311

A scalar tensor is defined as rank-0, size-1 tensor. PyTorch eager (mostly TensorIterator) supports device promotion of cpu scalar tensor, where you can have cross device tensors (cpu scalar tensor and cuda tensors) feeding to a single operator, and cpu scalar tensor would be promoted to a scalar.

We extended this support to nvfuser. A few changes that's required to support this:

API to query if a given tensor is indeed a scalar tensor is_scalar. Current criteria is tensor rank and size (utils.h & utils.cpp)
Update to partition logic where the device of a cpu scalar tensor is ignored. This should avoid us accidentally merging an operator of two cpu scalar tensors.
Integration code updated:
i. maps TS cpu scalar tensor into codegen scalar;
ii. skips usual tensor checks (vectorization / valid inputs) for cpu scalar tensor
iii. kernel arguments to extract scalar value from cpu scalar tensor
cpu scalar tests. Need to verify: 1. cpu scalar tensor with gpu tensor; 2. cpu scalar tensor with cpu scalar tensor; 3. cpu scalar tensor with cpu tensor; 4. cpu tensor with gpu scalar tensor
Note that, we briefly tried the alternative approach where we move cpu scalar tensor to gpu scalar tensor. Implementation is very straight forward, but a cuda tensor creation and copy is really slow. Hence the motivation to extract it into a scalar argument. More details in the issue #1311

jjsjann123 committed:
Merge remote-tracking branch 'csarofeen/devel' into HEAD

jjsjann123 committed:
Merge remote-tracking branch 'upstream/master' into HEAD

jjsjann123 committed:
Merge remote-tracking branch 'upstream/master' into upstream_master_bump_013122

rdspring1 committed:
Add tanh-backward support (#1420)

csarofeen committed:
Merge branch 'devel' of https://www.github.com/csarofeen/pytorch into upstream_master_bump_012622

csarofeen committed:
Merge pull request #1414 from csarofeen/upstream_master_bump_012622

Upstream master bump 012622

rdspring1 committed:
Set nondet_tol to 1e-5 for gradcheck in test_unary_ops (#1423)

* Set nondet_tol to 1e-5 for gradcheck
* Fix random seed for deterministic results

rdspring1 committed:
Add log-softmax, mean, var, and std operations (#1417)

* Implement operations to support AOTAutograd

jjsjann123 committed:
fixing stride order for expanded tensor (#71665) (#1431)

Summary:
The default initialization of stride order were not correct. This ended up with an expanded tensor showing wrong stride, since stride 0 is ignored by TensorIterator stride computation logic [Computing output strides].

Quick fix with cpp tests as well.

Note that things still look strange when we expand from a rank 1 size 1 tensor, as that gives us inconsistent strides.
```
In [7]: x = torch.rand([1])

In [8]: x.expand(1, 1, 4, 4).stride()
Out[8]: (0, 0, 0, 0)

In [9]: x.expand(4, 4, 1, 1).stride()
Out[9]: (0, 0, 1, 1)

In [10]: x.expand(4, 1, 4, 1).stride()
Out[10]: (0, 0, 0, 1)
```

Meanwhile, scalar tensor seems to work fine.
```
In [2]: x = torch.tensor(1.0)

In [3]: x.expand(4, 1, 1, 4).stride()
Out[3]: (0, 0, 0, 0)

In [4]: x.expand(4, 1, 4, 1).stride()
Out[4]: (0, 0, 0, 0)

In [5]: x.expand(4, 4, 1, 1).stride()
Out[5]: (0, 0, 0, 0)

In [6]: x.expand(1, 1, 4, 4).stride()
Out[6]: (0, 0, 0, 0)
```

Pull Request resolved: https://github.com/pytorch/pytorch/pull/71665

Reviewed By: mrshenli

Differential Revision: D33849958

Pulled By: davidberard98

fbshipit-source-id: 982cd7fa352747d1e094a022475d6d1381ba75e5
(cherry picked from commit 0e0b587fe18ed47f4e801bb55a10641b9decd6e4)

shmsong committed:
Additional type promotion tests involving cpu scalars (#1415)

* add category in bin op test

* format

* lint

* lint

* return early on tests that doesn't run

zasdfgbnm committed:
Add basic support for dtype ComplexFloat, ComplexDouble (#1427)

naoyam committed:
Cleans up some of the old IR transformation code with kir::ExprMutator (#1424)

csarofeen committed:
Avoid some unnecessary predicates. (#1429)

Co-authored-by: Naoya Maruyama <nmaruyama@nvidia.com>

naoyam committed:
Map everything between multiple outputs even for the CA Parallel Map (#1432)

eqy committed:
RMSNorm with tests/benchmarking (shapes based off of HuggingFace T5 on A100) (#1428)

naoyam committed:
Do not inline allocated scalars (#1434)

* Do not inline allocated scalars

naoyam committed:
print 0-dim tensors as tensors (#1442)

naoyam committed:
Index Hoisting (#1426)

* Hoist common index subexpressions

csarofeen committed:
Move welford to use nvfuser_index_t, pipe it through as a compile time type. (#1435)

csarofeen committed:
Use ParallelMap in expr sorting (#1436)

* Use parallel map in expression sorting.

naoyam committed:
Extend use of SimplyfingIrBuilder (#1448)

shmsong committed:
derive heuristics in intermediate reduction groups (#1447)

* derive heuristics in intermediate reduction groups

* unformat benchmarks

* lint

jjsjann123 committed:
View copy patch (#1450)

Fixes fusion guard build of view, which triggers assert failures in graph_fuser.cpp
```
        TORCH_INTERNAL_ASSERT(
            view->kind() == prim::view_copy ||
            view->kind() == prim::reshape_copy);
```
The old logic to find view_copy as a consumer of profiled constants are incorrect. We get lucky in our unit tests since we have two inputs both going to view_copy (tensor and shape) so the problem never arises.

jjsjann123 committed:
patching dtype for int32 and bool casting (#1449)

Fixes issues exposed in OpInfoTest pytorch#71299

This fixes a few op failures, which I didn't add comprehensive tests for. The idea is at some point we'll have OpInfoTests enabled and would run through those for extensive tests.

naoyam committed:
test fix (#1456)

Fixes #1455

csarofeen committed:
Fix input vectorization in the pointwise scheduler. (#1459)

csarofeen committed:
Rework vectorized load/stores. (#1457)

csarofeen committed:
Fix 896 and 1446 (#1461)

Fix sorting of tensor dims in reduction scheduler, fix binding size-0 inputs.

naoyam committed:
Refactoring of loop materialization (#1452)

* Refactoring of loop materialization

This is for promoting further index hoisting. Specifically, in some
cases loop index values are replaced at the time of final code
generation, which makes hoisting of expressions using those indices
rather annoyingly complex. Instead of doing so, this PR does indexing
with the final loop index values.

- Adds ForLoop::isTrivial to find if a loop does not need to be
materialized. For example, parallelized loops are not need to be
materialized with some exceptions.

- Uses isTrivial to set the loop index as its start value when indexing.
This makes some code in CudaCodeGenerator::handle(kir::ForLoop*)
unnecessary, and resulting code should be a little concise as there is
no "constexpr nvfuser_index_t i100 = 0;".

naoyam committed:
Index hoist follow-up (#1458)

* Relaxed index hoisting conditions

Indices with threadIdx/blockIdx are not previously hoisted when the
corresponding parallelized loops are are shared. This commits relaxes
the condition and allows for more cases of index hoisting.

Also, adds support of hoisting of unswitch predicates. Turned out it
just needed a trivial change.

naoyam committed:
Remove debug print (#1463)

jjsjann123 committed:
fixing parsing rule for empty reduction axes (#1454)

empty reduction axes in eager suggests full reduction on all dimensions. updated parsing rule to reflect that.

csarofeen committed:
Issue 1444 (#1462)

Avoid pass through arith functions, best effort support trivial inp->out sections in fusions.

jjsjann123 committed:
disabling 0-dim cuda tensor reduction/normalization (#1453)

1. Disables normalizatoin/reduction for 0-dim cuda inputs.
2. Patches fusion logic on reduction parsing rule

We are doing this because normalization for 0-dim behavior is tricky to justify (?!) and seems implementation specific. So we just leave this for eager.

naoyam committed:
Do not do special handling for broadcast domains in parallel dimension (#1464)

The CA Index map previously mapped broadcast and non-broadcast domains, so the extent with broadcast domains may not be the actual extent of the domain.

jjsjann123 committed:
Merge remote-tracking branch 'upstream/master' into HEAD

zasdfgbnm committed:
Type inference bug fix (#1443)

* Type inference bug fix

* format

* cleanup

* save

* format

* resolve review comments

zasdfgbnm committed:
Add complex scalar support (#1433)

jjsjann123 committed:
disabling reduction fusion on size-0 tensors (#1469)

Fixes #1466

Note that this is a WAR, not a proper fix on codegen to handle size-0

csarofeen committed:
Assert on zero size dimension. (#1470)

jjsjann123 committed:
Axes patch (#1476)

Existing logic access out of range dimension when target dimension is 0 (causing segfaults)

jjsjann123 committed:
patching aten failures on python tests (#1475)

Fixes failing test_binary_ops on pow by avoiding failing config

zasdfgbnm committed:
Fix type computation for complex abs (#1482)

* Fix type promotion for complex abs

* save

jjsjann123 committed:
Int bool tensor support (#1479)

Extended support for integer/boolean tensor.

csarofeen committed:
Fix RAW placement in outer most scope. (#1474)

naoyam committed:
Fixing codegen of fused warp reduce and broadcast (#1483)

zasdfgbnm committed:
Add tensor.view(dtype) overload support (#1481)

* Add tensor.view(dtype) overload support

* save

* format

* format

* clang-tidy

* Resolve review

* use reinterpret_cast

zasdfgbnm committed:
Enable some C++ tests for complex (#1472)

* Enable some tests for complex

* typo

* format

jjsjann123 committed:
Minor opinfo fixes (#1478)

1. disables fusion of bfloat tensors on pre-ampere device (causes nvrtc failure on inlined ptx assembly with bf, which is not supported on pre-ampere devices)
2. added type inference for rsub

jjsjann123 committed:
Casted alias (#1480)

supports alias on casted inputs. This allows us to handle fp16 running stats (fp16 norm layers).

Arguably we would want to do this in the parser code instead of inside Fusion::aliasOutputToInput. But that means we'll duplicate this code in a few places, which is ugly as well.

TODO:
properly handle segment io alias cpp test in PR #1471

zasdfgbnm committed:
Remove some NOLINTNEXTLINE (#1485)

* Remove some NOLINTNEXTLINE

* Remove some nolint

* save

jjsjann123 committed:
Squeeze scalar tensor (#1489)

Quick fix on squeeze of scalar tensor

jjsjann123 committed:
quick patch to drop aliased output from pushing to stack (#1471)

fix pytorch#67610
drop aliased output from pushing to stack

Currently we don't support to mark aliased output to be real outputs. which is tracked in #1488

jjsjann123 committed:
Merge remote-tracking branch 'upstream/master' into HEAD

jjsjann123 committed:
Debug improvements (#1328)

addding a few more debug dump and a quick doc helping people getting python repros;
removing obsolete code.

jjsjann123 committed:
Parser update on size 0 (#1490)

Disables fusion of empty tensor (numel == 0), which causes scheduler asserts with our latest scheduler update.

csarofeen committed:
Minor cleanup in runtime files. (#1465)

shmsong committed:
Fix vector reset for double buffered tensor on registers (#1491)

csarofeen committed:
WAR For Issue #1487 (#1492)

Avoid reduction/normalization scheduler if there's trivial reductions outside the reduction op.

csarofeen committed:
Minor fix for trivial reductions. (#1496)

* Minor fix for trivial reductions.


Co-authored-by: Naoya Maruyama <nmaruyama@nvidia.com>

naoyam committed:
Fix concrete domain selection with view rfactor domains (#1494)

rdspring1 committed:
Fixes to enable view fusion in LTC (#1451)

* Gather shape expressions for all sub-blocks in graph
* Add shape support for prim::view_copy and prim::reshape_copy
* Add FusionReductionViewShmoo test
* Set default scheduler to None
* Skip target expr with broadcast IDs for rfactor domains in BestEffortReplay
* Update MergeQueryFuncPtr to check if self tensor has shape information
* Remove traverseProfileIValues

jjsjann123 committed:
Merge remote-tracking branch 'upstream/master' into HEAD

jjsjann123 committed:
Merge remote-tracking branch 'origin/devel' into HEAD

jjsjann123 committed:
Reshape fix (#1499)

Fixes missing shape info for guard view.
The issues comes in that we are overwriting fusion_value_to_runtime_shape_ for each fusion group, instead of accumulating them. So later guard would have lost the entries needed.

csarofeen committed:
Merge pull request #1468 from csarofeen/upstream_master_bump_021622

Upstream master bump 021622

jjsjann123 committed:
adding native_batch_norm_backward into parser (#1501)

LTC request to support native_batch_norm_backward in parser

shmsong committed:
Minor fix on reference replay (#1505)

eqy committed:
InstanceNorm Channels Last 3D Benchmarks + InstanceNormBackward (#1438)

csarofeen committed:
Global memory communication (#1484)

naoyam committed:
Fused grid reduction and broadcast (#1495)

Co-authored-by: Christian Sarofeen <csarofeen@nvidia.com>

rdspring1 committed:
Change default gelu to use erf() instead of normcdf() (#1513)

* Change default Gelu to use erf instead of normcdf
* Remove Gelu unary op from helper.cu
* Define Gelu as composite operation
* Rename fast-gelu to tanh-gelu

naoyam committed:
Print strides of input and output tensors (#1520)

jjsjann123 committed:
fixing contiguous on broadcasted dimension (#1519)

Fixes #1514

Issue arises when we try to collapse a dimension to a broadcasted dimension. PyTorch mark all dimension with stride 1 to be contiguous. This used to be fine until recently we changed it that broadcasted dimension with stride 0 can now be put as a faster dimension. So now we can be collapsing a normal dimension into a broadcasted dimension and our index is wrong, since broadcasted dimension has stride 0.

The solution is to explicitly check inner dimension stride for contiguous dimension and only collapse when inner dimension is not broadcasted.

naoyam committed:
Disallow passing immediate scalars as inputs (#1522)

kevinstephano committed:
Fixing test_linear to use a 3-D input tensor instead of 2-D one.  The 2-D tensor was triggering eager mode bias fusion that impacted TF32 math on Ampere. (#1515)

rdspring1 committed:
Select output reference tensor correctly for pointwise scheduler (#1500)

* Pointwise scheduler heuristics requires a minimum number of axes.
* The output reference tensor should respect this requirement.
* Only use valid reference tensors during pointwise heuristics analysis

csarofeen committed:
Make sure WAR syncs for gather ops are properly generated. (#1516)

Co-authored-by: Naoya Maruyama <nmaruyama@nvidia.com>

naoyam committed:
Relaxing validation of aligned vectorization (part 1) (#1508)

naoyam committed:
Disable index hoisting with PYTORCH_NVFUSER_DISABLE_INDEX_HOIST (#1526)

naoyam committed:
Fixes #1523 (#1525)

shmsong committed:
Mma operator and volta mma integration (#1439)

* initial volta support

* mma parallel type && cleanup

* cleanup

* alignment

* comment

* change request

* fix same parallel type

* move validation pass

* comment and cleanup

* lint

* comment and cleanup

* comment and format

rdspring1 committed:
Fix Issue 1502 (#1503)

* Propagate new symbol throughout fusion using ValReplacementMutator
* Replace FusionViewFailPersistent with FusionViewPersistentShmoo
* Create separate test-gpu-view.cpp for view tests
* Move replaceValue to ir_utils

naoyam committed:
Drop mappings of unused reference IDs (#1527)

csarofeen committed:
Add contiguous indexing through broadcast roots, fix indexing to support it. (#1528)

naoyam committed:
Add fusion_args option for PYTORCH_NVFUSER_DUMP (#1532)

fusion_args prints arguments given to runFusion.

kernel_args prints arguments given to generated CUDA kernels

naoyam committed:
Fixes #1530 (#1531)

naoyam committed:
Fixes #1497 (#1533)

naoyam committed:
Fixes validation of vectorization with contig indexing (#1537)

* Fixes validation of vectorization with contig indexing

True contig indexing needs reference tensors, so finding vectorized
contig domains at the initial validation time can result in false
positives and negatives. Fixed by filling that information at the time
of indexing.

Also considered to keep it separated from indexing and fill it at the
validation time, but it would end up replicating the same logic as reference
replay.

Closes #1534

naoyam committed:
Do not propagate rfactor flag (#1542)

To highlight the impact of the change, renamed `IterDomain::clone()` to `IterDomain::cloneWithoutRFactor()`.

csarofeen committed:
Rework compute at map concrete ID resolution for view (#1535)

csarofeen committed:
Makes the reference tensor have the complete history of transformations V2 (#1536)

naoyam committed:
Fix ID counting with rfactor (#1544)

zasdfgbnm committed:
flatten support - C++ only (#1545)

* save

* save

* save

* save

naoyam committed:
Fix ComputeAtRootDomainMap with broadcast in view root domains (#1552)

* Fix ComputeAtRootDomainMap with broadcast in view root domains

Fixes #1549

naoyam committed:
Enable the pointwise scheduler to vectorize contig merged domains (#1529)

* Allow vectorization with contig-merged domains in pwise scheduler

naoyam committed:
Forward merging of trivial-reduction dims in producers (#1553)

* Forward merging of trivial-reduction dims in producers

* Enable trivial reduction forwarding only when trivial reduction domain
is a root domain.

For example, splitting a reduction domain by 1 and merging it with
another non-reduction domain would result in a trivial-reduction merge.

Probably possible to allow such non-root trivial reduction domains, but that
would mean, e.g., a leaf domain would be mappable yet its root domain
could be unmappable, which seems rather confusing. Considering such
transformations would be unlikely, not enabling forwarding would be fine
and would cause less surprise.

naoyam committed:
Propagate root domain mappings from rfactor to root domains (#1556)

* Propagate root domain mappings from rfactor to root domains in
ComputeAtRootDomainMap

The main purpose of ComputeAtRootDomainMap is to find unmappable domains
for comptueAt. This analyais is done by traversing a fusion in a
backward direction. Currently, the traversal only visits arithmetic
expressions, so information propagation is done from consumer tensors to
producer tensors. This propagation is also required from rfactor domains
to root domains. Previously it doesn't really matter as rfactor is
limited reduction domains, but that's not the case with view.

This change also means that ComputeAtRootDomain does not guarantee
one-to-one mappings. For example,

```
tv0: [I0, I1]
tv1 = view(tv0); // tv1: [I0*I1/N, N]
```

I.e., the view op is done first merging the two domains of `tv0` and
then splitting it by N. Note that both of the two rfactor axes of `tv1`
are now mapped with the two axes of `tv0`.

Because of this change, `ComputeAtRootDomainMap:mapBestEffort` and other
mapping functions between a producer and a consumer that is supposed to
return a one-to-one map can fail.
`ComputeAtRootDomainMap::getMappableDims` is fine as it just grabs any
domain that is mappable.

`ComputeAtRootDomainMap::mapConsumerToProducer` and
`ComputeAtRootDomainMap::mapProducerToConsumer` were used in
`TransformReplay::replayPasC` and `TransformReplay::replayCasP`, but
they don't really need to use `ComputeAtRootDomainMap` but just
`PairwiseRootDomainMap` is sufficient, so replaed the usages with the
pairwise variant.

shmsong committed:
Python test fix (#1558)

* Minor fix on python test

zasdfgbnm committed:
flatten support - python (#1550)

Add flatten support on the python side

naoyam committed:
remove debug printing (#1560)

kevinstephano committed:
Added a more helpful error message when checking for empty outputs on… (#1559)

* Added a more helpful error message when checking for empty outputs on the Fusion.

* Clang fix.

shmsong committed:
Fix segmenter issue with unary op forwarding and debug print (#1561)

* do not re-compute unary op with output and allow expr duplication in debug print.

shmsong committed:
shared memory allocation update (#1498)

* always allocate dynamic smem

* add driver API call for large smem usage

Co-authored-by: Christian Sarofeen <csarofeen@nvidia.com>

jjsjann123 committed:
Merge commit '873ced7cd02f3f33e65b2020fd52801df963cd25' into HEAD

jjsjann123 committed:
Merge remote-tracking branch 'upstream/master' into upstream_merge_april

jjsjann123 committed:
Revert "disabling view"

This reverts commit 2d5e4cff854749da7e5d83ac3eac60c80351d973.

jjsjann123 committed:
Merge remote-tracking branch 'origin/devel' into upstream_master_bump_040722

jjsjann123 committed:
clang-tidy flake8 ci_workflows
quick test fix

csarofeen committed:
Merge pull request #1563 from csarofeen/upstream_master_bump_040722

Upstream merge april

naoyam committed:
Cleaning up ReductionOp (#1565)

* Cleaning up ReductionOp

- GridReduction inherits ReductionOp. A grid reduction is now
represented with just a single GridReduction expression (previously it
was with ReductionOp AND GridReduction)
- The runtime gridReduction code is extended to do block reduction as
well. This simplifies code generation as it does not need to generate
a separate block-reduction call.

rdspring1 committed:
Refactor TensorViewBuilder helper functions in Benchmarks (#1557)

 * Move makeSymbolicTensor, makeContigTensor, makeContigConcreteTensor, and makeConcreteTensor to utils.h
 * Remove duplicate definitions
 * Add initial transpose benchmarks

jjsjann123 committed:
fixing clang-tidy (#1566)

Fixes clang-tidy issue that's complaining about goto in ASSERT_ANY_THROW

naoyam committed:
clang issues lots of warnings of comparisons in the C++ test files. (#1567)

Likely because now a different compilation option is used as they were moved from
the test directory?

rdspring1 committed:
Rename cache functions to camelCase format (#1569)

 * cache_after to cacheAfter
 * cache_before to cacheBefore
 * cache_fork to cacheFork

jjsjann123 committed:
turn on nhwc for more PW ops (#1570)

Turning permutatoin support on for a few pointwise ops:
aten::softplus; aten::threshold; aten::where; aten::lerp

Initial request coming from an example with poor perf for channels last input tensor.

```
def smelu(x: torch.Tensor, beta: float):
    zero=torch.empty((), dtype=x.dtype, device=x.device).fill_(0.) # to avoid synchronizing h2d transfer
    return torch.where(x >= beta, x, torch.where(x <= - beta, zero, (x+beta)*(x+beta)/(4*beta)))

memory_format=torch.channels_last
x = torch.randn(10240000, device="cuda").view(128, 80, 10, 100).to(memory_format=memory_format).detach().requires_grad_(False)


scripted = torch.jit.script(smelu)
smelu(x, 2.)
gO = torch.rand_like(x)
with torch.jit.fuser("fuser2"):
    for _ in range(10):
        x.grad = None
        out = scripted(x, 2.)
```

jjsjann123 committed:
consistent clamp for reversed min/max bound (#1572)

Fixes pytorch#75558 (comment)

Updated clamp logic to be consistent with aten. This avoids us producing different result when clamp was given a min/max argument where min > max.
We don't have an issue opened for this. It is also tricky to argue the right behavior, but getting better consistency with eager is always(?!) a good thing.

naoyam committed:
clang-format (#1581)

naoyam committed:
Ensure WelfordOp always have non-null var input (#1583)

naoyam committed:
Add PYTORCH_NVFUSER_DISABLE_PREDICATE_ELIMINATION (#1574)

If defined, disable predicate elimination. Mma tests would fail as
predicates are not (currently) supported with Mma.

naoyam committed:
Fix issue #1579 (#1580)

jjsjann123 committed:
patching lint action repo (#1585)

jjsjann123 committed:
Disable fusion on high rank tensor (#1573)

Disables fusion on tensor with input rank > 8, which causes assert in a few places where we have limited tensor rank support.

This avoids going through fallback on a fusion graph which we can't handle. Issue raised by @ngimel via slack discussion.

naoyam committed:
Temporarily disable a predicate elimination test. (#1591)

Fails due to #1571

jjsjann123 committed:
[JIT] nvfuser test - use only major & minor versions (#1587)

torch.version.cuda.split('.') can have 2 or 3 elements depending on whether the version string contains a patch number. This updates the test so it doesn't error out when the version has > 2 parts.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/75706

Approved by: https://github.com/ngimel

Co-authored-by: David Berard <dberard@fb.com>

jjsjann123 committed:
[NVFuser] fix incorrect vector size in chunk move (#1589)

Internally there's some modes that run with -D_GLIBCXX_ASSERTIONS, which was throwing out of bounds errors at `strides[output_index]`

Pull Request resolved: https://github.com/pytorch/pytorch/pull/75765

Approved by: https://github.com/jjsjann123, https://github.com/cpuhrsch

Co-authored-by: David Berard <dberard@fb.com>

jjsjann123 committed:
[NVFuser] make comparators obey strict weak ordering (#1590)

Internally there is an at-runtime assert that was failing. The assert is checking to make sure that comparators obey strict weak ordering
https://en.cppreference.com/w/cpp/named_req/Compare

In particular `comp(a, a) == false` was failing.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/75713

Approved by: https://github.com/jjsjann123, https://github.com/csarofeen

Co-authored-by: David Berard <dberard@fb.com>

jjsjann123 committed:
baby steps on patching inf/nan behavior & aten::amin support in nvfuser (#1588)

Fixes #75622

1. Instead of getting max/min_value for reduction init value, we go with (-)infinity instead so we can properly preserve inf inputs;
2. Adding inf/(-)inf/nan for float value.
3. Adding aten::amin in nvfuser (@kevinstephano @rdspring1 for review)
Pull Request resolved: https://github.com/pytorch/pytorch/pull/75646
Approved by: https://github.com/rdspring1, https://github.com/kevinstephano, https://github.com/ngimel

jjsjann123 committed:
patching clamp for one sided clamp (#1586)

Fixes #75088

The solution is just to avoid putting random value for non-specified clamp as pointed out in https://github.com/pytorch/pytorch/issues/75088#issuecomment-1093410036

Pull Request resolved: https://github.com/pytorch/pytorch/pull/75558
Approved by: https://github.com/ngimel

jjsjann123 committed:
Device constant (#1593)

Fixes #1331

Added an entry for device constant

naoyam committed:
Fix predicate elimination (#1582)

* Fix predicate elimination

Closes #1571

csarofeen committed:
IterDomain Graphs/Map for Compute At Refactor (#1548)

Make a new mechanism called iter domain graphs to easily traverse through Exprs between TVs and IDs to understand how IterDomains map.

csarofeen committed:
Fix sync placement in some cases where it was less than optimal or wrong. (#1600)

* Fix placment of added test.

* Place RAW sync at computeAt position rather than unroll position

When a shared-mem tensor is unrolled, which shouldn't be common as
unroll is meant to allocate enough registers for loop unrolling, its RAW
sync is needed at the computeAt loop as there are consumers sharing the
computeAt loop.


Co-authored-by: Naoya Maruyama <nmaruyama@nvidia.com>

csarofeen committed:
Unify disjoint set implementations. (#1596)

* Unify disjoint set implementations.

Co-authored-by: Naoya Maruyama <nmaruyama@nvidia.com>

naoyam committed:
Remove unused class (#1603)

naoyam committed:
Validate and parallelize sibling tensors (#1602)

Always group sibling ID into the same disjoint set in the Loop map
Validation sibling tensors are strictly same transformations

naoyam committed:
Move and cleanup predicate elimination (#1605)

rdspring1 committed:
remove print statement (#1607)

naoyam committed:
clang-format (#1611)

rdspring1 committed:
Add FusionViewConcreteDomain tests (#1609)

naoyam committed:
Fix error check (#1616)

rdspring1 committed:
Remove assertion (#1617)

naoyam committed:
Consumer dependent reference (#1615)

* Create a reference tensor so that it follows the same view path as the
indexed or predicated tensor

shmsong committed:
Quick fix on serial shared mem writes (#1604)

* repro and fix serial smem write with parallel smem read.

shmsong committed:
Minor fix on sync insertion (#1618)

fix insertion point when 2 shared mem producers on different loop nest level.

jjsjann123 committed:
[NVFuser] don't decompose linear if we don't have shape info (#1592)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/75770

Approved by: https://github.com/jjsjann123, https://github.com/robieta

Co-authored-by: David Berard <dberard@fb.com>

zasdfgbnm committed:
Simplify TransformPropagator::TransformPropagator (#1620)

Looks like this method could be simplified: First, there is no need to separate forward vs backward. Second, we could use a `std::unordered_set` in addition to the `std::deque` to avoid inserting redundant computation.

kevinstephano committed:
Add a matching lerp implementation to eager mode. (#1612)

Fixes #1597 
Add a matching lerp to eager mode.

Co-authored-by: jjsjann123 <alex.jann2012@gmail.com>
Co-authored-by: jjsjann123 <jiej@nvidia.com>

zasdfgbnm committed:
Add view_as_real support (#1512)

I added two new types `Float_2` and `Double_2` to `DataType`. In the future, we might need to add more vector types if we decide to support full `view(dtype)`.

I also renamed `ViewDtypeOp` into `ViewAsScalar` and changed its definition: now `ViewAsScalar` has one input TV which must have a vector dtype, and this operator will append a new `IterDomain` of type `IterType::VectorComponent` which iter through the vector dimension. For example, if the input TV of `ViewAsScalar` is `TV[iS{i0}]` of type `Float_2`, then the output will be `TV[iS{i0}, vS{2}]` of type float.

`EraseType` is also renamed to `BitCast`.

I also added lowering for `ViewAsScalar` in `IndexLowering`.

csarofeen committed:
Add channels last, ResNet, and ResNext sizes for Batch Norm benchmarks. (#1622)

naoyam committed:
Exact root domain mapping (#1624)

* Add ExactRootDomainMap

Broadcast domains are not mapped with non-broadcast domains

shmsong committed:
minor fixes on expr sorting and insert syncs (#1625)

* deterministic merge order in expr sorting
* discard smem writes that do not need sync

jjsjann123 committed:
Permutation extended (#1614)

Extended permutation support in integration (See more details on #1601). This update allows us to better support permutation propagation on tensors, specifically for binary ops with inputs of different ranks. Our goal is to avoid permuting tensors unless absolutely necessary. We try to preserve the permutation propagation rule in aten, with some known limitation at the time.

The idea in this implementation is the same as with our existing code, which is to permute input/output tensors outside of codegen: For a simplified binary op scenario: `output = binaryOp(input0, input1)`

1. In a simple case where `input0` and `input1` come with the same rank & permutation order, our output would preserve the same permutation;
2. For cases where `input0` and `input1` come with different ranks but with **compatible** permutation, the tensor with the higher rank dictates the permutation of the output;
3. For cases where `input0` and `input1` come with different ranks but with **in-compatible** permutation, this is where permutation propagation fails and the output tensor will be contiguous.

By **compatible** permutation, it means that we can permute the higher rank tensor to contiguous format, and then apply a second permutation to the tensor with lower rank to match their axes. This check is implemented in `MemoryFormat::broadcastToRank(int lower_rank)`.

Some concrete example (note that we comply with eager propagation on cases 1-3, but diverge in behavior for cases 4, 5):
1. different rank & same permutation
```
    t0 = torch.randn(b, h, w, c).cuda().permute([0, 3, 1, 2])  # stride (hwc, 1, wc, c)
    t1 = torch.randn(h, w, c).cuda().permute([2, 0, 1])  # stride (1, wc, c)
    out = scripted_add(t0, t1)  # stride (hwc, 1, wc, c) preserving memory format of t0
```
2. different rank & compatible permutation
```
    t0 = torch.randn(b, h, w, c).cuda().permute([0, 3, 1, 2])  # stride (hwc, 1, wc, c)
    t1 = torch.randn(c, h, w).cuda()  # stride (hw, w, 1)
    out = scripted_add(t0, t1)  # stride (hwc, 1, wc, c) preserving memory format of t0
```
3. different rank & compatible permutation with broadcasting
```
    t0 = torch.randn(b, h, w, c).cuda().permute([0, 3, 1, 2])  # stride (hwc, 1, wc, c)
    t1 = torch.randn(c).cuda().unsqueeze(-1).unsqueeze(-1)  # stride (1, 1, 1)
    out = scripted_add(t0, t1)  # stride (hwc, 1, wc, c) preserving memory format of t0
```
4. different rank & in-compatible permutation
```
    t0 = torch.randn(b, h, w, c).cuda().permute([0, 3, 1, 2])  # stride (hwc, 1, wc, c)
    t1 = torch.randn(h, w).cuda()  # stride (w, 1)
    jit_out = scripted_add(t0, t1)  # stride (hwc, 1, wc, c)  # stride (hwc, wc, c, 1)  # nvfuser outputs contiguous tensor
    eager_out = eager_add(t0, t1)  # stride (hwc, 1, wc, c)  # stride (hwc, 1, wc, c)  # TI preserves memory format of LHS operand
```
5. different rank & in-compatible permutation
```
    t0 = torch.randn(c, h, w).cuda()  # stride (hw, w, 1)
    t1 = torch.randn(b, h, w, c).cuda().permute([0, 3, 1, 2])  # stride (hwc, 1, wc, c)
    jit_out = scripted_add(t0, t1)  # stride (hwc, 1, wc, c)  # stride (hwc, 1, wc, c)  # nvfuser preserves memory format of highest rank tensors
    eager_out = eager_add(t0, t1)  # stride (hwc, 1, wc, c)  # stride (hwc, hw, w, 1)  # TensorIterator preserves memory format of LHS operand
```

naoyam committed:
cleanup (#1627)

shmsong committed:
[Fusion Segmenter QOL] add segmenter logging and improve debug print (#1634)

* segmented fusion debug print
* segmenter logging

naoyam committed:
Preliminary support of grouped reduction (#1623)

* Grouped reduction

* PR feedback

* Consolidate Welford and GroupedReduction rfactor

csarofeen committed:
Move vectorize helpers out of pointwise scheduler to someplace common. (#1629)

* Vectorize runtime cleanup (#1635)
* Add expanded vectorize to Reduction and Normalization schedulers (#1637)
* Make sure we check the right index type size in schedulers for welford.
* PR review collection and cleaning up the vectorize PR stack (#1638)

jjsjann123 committed:
breaking stuff (#1636)

update CI tests with lintrunner to ensure format stuff are running in CI

jjsjann123 committed:
Merge remote-tracking branch 'upstream/master' into HEAD

csarofeen committed:
Minor benchmark fixes. (#1639)

zasdfgbnm committed:
Update obselete comment (#1646)

These mapping modes are renamed, and I checked `lower_expr_sort.cpp` quickly, they are also using `IdMappingMode::LOOP`.

zasdfgbnm committed:
nit cleanup in lower_loops.cpp (#1648)

jjsjann123 committed:
Merge remote-tracking branch 'origin/master' into HEAD

shmsong committed:
remove v3 option (#1647)

jjsjann123 committed:
fixing workflow

naoyam committed:
Fix grid sync allocation and codegen (#1650)

naoyam committed:
Grouping allreduces (#1644)

Rename is_fused_ to is_allreduce_
Add a C++ test based on batchnorm backward

csarofeen committed:
Merge pull request #1649 from csarofeen/master_bump_0501

Merge upstream.

shmsong committed:
Quick patch on pred removal pass (#1642)

jjsjann123 committed:
removing workflow pull (#1654)

naoyam committed:
Print ca map (#1653)

* Add a debug dump option to print the computeAt map
(PYTORCH_NVFUSER_DUMP=ca_map)

csarofeen committed:
Add returns and indenting to LHS of broadcast and reduction ops. (#1658)

csarofeen committed:
Emit thread predicates for vectorized instructions. (#1660)

naoyam committed:
Consolidate all the PYTORCH_NVFUSER_DISABLE_ options like PYTORCH_NVFUSER_DUMP (#1666)

* Consolidate all the PYTORCH_NVFUSER_DISABLE_ options like
PYTORCH_NVFUSER_DUMP

csarofeen committed:
Avoid redundant writes on global memory. (#1661)

csarofeen committed:
Cleanup heuristic structs. (#1663)

csarofeen committed:
Cleanup some warnings. (#1669)

shmsong committed:
Code clean up and minor fix on dynamic shared alignment (#1670)

csarofeen committed:
Add some TIMM micro-benchmarks and some minor benchmark fixes (#1656)

Fix benchmarks.../utils.h path in headers
Fix printing schedule parameters when fusion segmeneter runs but still only makes one kernel

csarofeen committed:
Make sure transform propagate always takes the most replayed replay. (#1662)

csarofeen committed:
Small IrCloner fix (#1668)

csarofeen committed:
Add debug option PYTORCH_NVFUSER_DUMP=perf_debug_verbose (#1657)

At the time of running a fusion/segment it prints:

1. Where the segment begins execution, where it ends execution
2. FusionIR associated with the operation
3. All input tensor metadata
4. Compiler log with register usage
5. Scheduler parameters
6. Launch Parameters
7. Bytes processed by fusion
8. Kernel execution time
9. Effective GB/s

csarofeen committed:
Make sure vectorization and inlining is consistent in the reduction/norm kernels. (#1665)

csarofeen committed:
Simultaneous cleanup conflicting. (#1672)

csarofeen committed:
Rework pwise scheduler for better broadcasting perf on TIMM VIT. (#1664)

shmsong committed:
Use fp64 reference for bfloat16 unary op tests (#1674)

jjsjann123 committed:
Memory format propagation parser patch (#1675)

mark memory format on a few ops that does support memory format propagation (clamp/softplus/threshold)
This causes us to output tensor in wrong shape, causing failure on torchvision models pytorch#76791

jjsjann123 committed:
Profile ivalue multiple (#1677)

Fixes profiling asserts on seeing multiple inconsistent profile ivalues.

Instead of asserting, we now just wipe out profile_ivalues and putting a debug attribute in the profile node profile_node->s_(Symbol::attr("profile_failed"), "varying profile values");. So during fusion pass, we'll just not have profiled constant for those values and might miss out some fusion opportunity.

Alternatively, we should just be able to ignore inconsistent ivalues and expect our guard/fallback to work at runtime. But that's too aggressive and we'd like to avoid frequent fallback.

This also fixes the part of failure mentioned pytorch#76791

naoyam committed:
Use int rather than int64_t for hoisted index values when the kernel (#1685)

index type is int.

This feels rather hacky.

zasdfgbnm committed:
Use PYTORCH_NVFUSER_DISABLE in tests (#1688)

shmsong committed:
Patch sync insertion for redundant predicated writes (#1684)

* patch sync insertion

jjsjann123 committed:
Opinfo patch (#1682)

1. Fixes some opinfo tests on __half math for log_softmax;
2. Update parser to avoid fusing nodes with explicit reduced precision math type.

TODO: Expand the parser update to other ops.

csarofeen committed:
Only extract bias out of linear layer if input ndims >= 4 (#1679)

restrict decomposition of linear for input.rank() >=4
updating python tests

Co-authored-by: jjsjann123 <jiej@nvidia.com>

naoyam committed:
Fix printing of vector (#1693)

* Use tellp not tellg

naoyam committed:
Fix usage of ArrayRef (#1697)

jjsjann123 committed:
adding permutation for dropout (#1691)

Turns on permutation support for randLike ops;
Fixing permutation support for softmax.

jjsjann123 committed:
Merge remote-tracking branch 'upstream/master' into HEAD

jjsjann123 commented:
cc'ing @davidberard98 
Don't think we can get nightlies to run on this. I'll mark it as ready for review once my local tests finished.
Likely would get some style changes as well. :crossed_fingers:

jjsjann123 commented on aten/src/ATen/core/interned_strings.h:
```diff
@@ -51,6 +51,7 @@ namespace c10 {
   _(prim, reshape_copy)              \
   _(prim, squeeze_copy)              \
   _(prim, unsqueeze_copy)            \
+  _(prim, flatten_copy)              \
```
This should be the only change that's outside of nvfuser code base.

davidberard98 commented:
Marking as "ready for review" so I can import it.

jjsjann123 committed:
clean up tests skip for alias

guarding cuFuncGetAttribute/SetAttribute on ROCm

HIP fixes

fixing win build

clang-format

jjsjann123 committed:
clean up tests skip for alias

jjsjann123 committed:
Merge remote-tracking branch 'origin/viable/strict' into HEAD

jjsjann123 committed:
guarding cuFuncGetAttribute/SetAttribute on ROCm

malfet commented:
Just curious: considering that commits are cheap in GHF world: is there really a benefit in such aggregate commits?
Obvious drawbacks: 
- Regression from one particular improvements results in revert of the entire change
- it's hard for community developers to contribute to NVFuser
- Large PRs are harder to keep uptodate with trunk
- It's hard to populate changelog
- Individual contributions are lost in this bulk commit

jjsjann123 committed:
HIP fixes

naoyam committed:
Ignore reduction axes when unsqueezing (#1703)

* Ignore reduction axes when unsqueezing

* Similar fix for squeeze

jjsjann123 committed:
Squeeze zero patch (#1704)

Fixes recently test_squeeze_zero breakage, copy squeeze op on scalar tensor

naoyam committed:
Fix un-welford transformation (#1705)

* Fix un-welford transformation

naoyam committed:
Re-entrant grid reductions and outer reduction performance (#1698)

Allow re-entrant grid reductions, allowing them to be placed inside loops without turning the kernel cooperative.
Rework outer reduction performance.
Co-authored-by: Christian Sarofeen <csarofeen@nvidia.com>

naoyam committed:
Fix predicate with contig merge and broadcast (#1707)

csarofeen committed:
Add another TIMM micro-benchmark. (#1680)

jjsjann123 committed:
fixing win build

jjsjann123 committed:
Merge remote-tracking branch 'origin/viable/strict' into HEAD

jjsjann123 committed:
clang-format

jjsjann123 committed:
Merge remote-tracking branch 'origin/viable/strict' into HEAD

seemethere commented:
> Just curious: considering that commits are cheap in GHF world: is there really a benefit in such aggregate commits? Obvious drawbacks:
> 
> * Regression from one particular improvements results in revert of the entire change
> * it's hard for community developers to contribute to NVFuser
> * Large PRs are harder to keep uptodate with trunk
> * It's hard to populate changelog
> * Individual contributions are lost in this bulk commit

+1 to this, I see there are commits dating all the way back to December in this pull request.

Maintaining a branch this big must be difficult to do, is there not a way we can make this easier for you to contribute these changes without maintaining such a large divergence?

seemethere reviewed (changes requested):
Considering we have a branch cut for 1.12 coming soon and a PR of this size could introduce a lot of uncertainty for having a stable release branch I say we should hold off on merging this PR until after the branch cut occurs and additional testing can be done within nightlies.

davidberard98 commented:
@seemethere these are patches required to re-enable nvfuser by default, which is planned for 1.12.

seemethere commented:
> @seemethere these are patches required to re-enable nvfuser by default, which is planned for 1.12.
> 
> cc @atalman to verify that this is okay

We probably won't even have time to verify that these changes are okay in the nightlies prior to the branch cut though so I don't see how we can plan to enable it by default in that time frame.

davidberard98 commented:
@jjsjann123 is it possible to get a smaller PR to fix to the torchvision issue instead

davidberard98 commented:
@seemethere are you suggesting we land this after 5/19 and then cherry-pick into 1.12 if nightlies are okay, or to postpone it entirely?

eellison commented:
@seemethere I understand this is a big pr to cherry-pick, but we're still ahead of the branch cut. Imo it's also less risky because there is a well-defined fallback path - switch default enablement to NNC.

jjsjann123 commented:
> @jjsjann123 is it possible to get a smaller PR to fix to the torchvision issue instead

This PR targets more than just patching the torchvision issue. Perf tuning on kernels in this devel update is also an important target for our 1.12 enablement. :pray:

jjsjann123 committed:
fixing lambda capture for MSVC

seemethere commented:
> @seemethere are you suggesting we land this after 5/19 and then cherry-pick into 1.12 if nightlies are okay, or to postpone it entirely?

I'd say _any_ PR >10k LOC shouldn't be merged into master so close to a branch cut or cherry picked into a release branch. This doesn't just apply to this particular PR.

jjsjann123 committed:
Merge remote-tracking branch 'csarofeen/upstream_master_bump_0517' into HEAD

malfet commented:
@eellison please do not use githubbot to merge this change but rather let it go thru internal CI/CD to get the full signal.

eellison commented:
> Merge failed due to list index out of range
Raised by https://github.com/pytorch/pytorch/actions/runs/2347398630

cc @seemethere, do you know anything about this ? it landed internally

eellison commented:
```
  File ".github/scripts/trymerge.py", line 923, in main
    pr.merge_into(repo, dry_run=args.dry_run, force=args.force, comment_id=args.comment_id)
  File ".github/scripts/trymerge.py", line 695, in merge_into
    repo._run_git("commit", f"--author=\"{self.get_author()}\"", "-m", msg)
  File ".github/scripts/trymerge.py", line 571, in get_author
    authors = self.get_authors()
  File ".github/scripts/trymerge.py", line 566, in get_authors
    rc[self.get_committer_login(idx)] = self.get_committer_author(idx)
  File ".github/scripts/trymerge.py", line 527, in get_committer_author
    return self._fetch_authors()[num][1]
```

seemethere commented:
@pytorchbot force merge this

seemethere commented:
Looks like this PR is so massive that our GHF tooling can't actually handle it correctly, investigating on how to unblock

seemethere commented:
Merged manually with the following modifications to `trymerge.py`:

```diff
diff --git a/.github/scripts/trymerge.py b/.github/scripts/trymerge.py
index f482358a85..067ca14184 100755
--- a/.github/scripts/trymerge.py
+++ b/.github/scripts/trymerge.py
@@ -521,10 +521,12 @@ class GitHubPR:
         return authors

     def get_committer_login(self, num: int = 0) -> str:
+        return "jjsjann123"
         return self._fetch_authors()[num][0]

     def get_committer_author(self, num: int = 0) -> str:
-        return self._fetch_authors()[num][1]
+        return "jjsjann123 <jiej@nvidia.com>"
+        # return self._fetch_authors()[num][1]

     def get_checkrun_conclusions(self) -> Dict[str, str]:
         """ Returns list of checkrun / conclusions """
```

seemethere commented:
This is the _last_ time we will be merging a PR of 20k lines
